### ISSUE TITLE
Deflation output changes v3.1.2 vs v3.2.2 - 3.2.2 no longer deflates coords

### ISSUE THREAD

**Issue by durack1** [PCMDI/cmor#158]
I've just seen the `areacello` file increase from 41.4Kb (3.1.2) to 62.3Kb (3.2.2) with no change in the arguments given to `cmor.set_deflate(varid,1,1,1)`.

The ncdump output of the v3.1.2 file (41.4Kb):
[areacello_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn-CMOR312.txt](https://github.com/PCMDI/cmor/files/923227/areacello_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn-CMOR312.txt)
```
netcdf areacello_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn {                                                                    
dimensions:                                                                                                                             
        lat = 180 ;                                                                                                                     
        lon = 360 ;                                                                                                                     
        bnds = 2 ;                                                                                                                      
variables:                                                                                                                              
        double lat(lat) ;                                                                                                               
...                                                                                      
                lat:_Storage = "chunked" ;                                                                                              
                lat:_ChunkSizes = 180 ;                                                                                                 
                lat:_DeflateLevel = 1 ;                                                                                                 
                lat:_Shuffle = "true" ;                                                                                                 
                lat:_Endianness = "little" ;                                                                                            
        double lat_bnds(lat, bnds) ;                                                                                                    
                lat_bnds:_Storage = "chunked" ;                                                                                         
                lat_bnds:_ChunkSizes = 180, 2 ;                                                                                         
                lat_bnds:_DeflateLevel = 1 ;                                                                                            
                lat_bnds:_Shuffle = "true" ;                                                                                            
                lat_bnds:_Endianness = "little" ;                                                                                       
        double lon(lon) ;                                                                                                               
...                                                                                   
                lon:_Storage = "chunked" ;                                                                                              
                lon:_ChunkSizes = 360 ;                                                                                                 
                lon:_DeflateLevel = 1 ;                                                                                                 
                lon:_Shuffle = "true" ;                                                                                                 
                lon:_Endianness = "little" ;                                                                                            
        double lon_bnds(lon, bnds) ;                                                                                                    
                lon_bnds:_Storage = "chunked" ;                                                                                         
                lon_bnds:_ChunkSizes = 360, 2 ;                                                                                         
                lon_bnds:_DeflateLevel = 1 ;                                                                                            
                lon_bnds:_Shuffle = "true" ;                                                                                            
                lon_bnds:_Endianness = "little" ;
        float areacello(lat, lon) ;
...
                areacello:_FillValue = 1.e+20f ;
                areacello:_Storage = "chunked" ;
                areacello:_ChunkSizes = 180, 360 ;
                areacello:_DeflateLevel = 1 ;
                areacello:_Shuffle = "true" ;
                areacello:_Endianness = "little" ;

// global attributes:
...
                :cmor_version = "3.1.2" ;
                :_SuperblockVersion = 2 ;
                :_IsNetcdf4 = 1 ;
                :_Format = "netCDF-4 classic model" ;
}
```
The ncdump output of the v3.2.2 file (62.3Kb) which seems to no longer be deflating the coordinates, and shows a different value for `_SuperblockVersion`:
[areacello_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn-CMOR322.txt](https://github.com/PCMDI/cmor/files/923228/areacello_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn-CMOR322.txt)
```
netcdf areacello_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn {                                                                   
dimensions:                                                                                                                            
        lat = 180 ;                                                                                                                    
        lon = 360 ;                                                                                                                    
        bnds = 2 ;                                                                                                                     
variables:                                                                                                                             
        double lat(lat) ;                                                                                                              
...                                                                                      
                lat:_Storage = "contiguous" ;                                                                                          
                lat:_Endianness = "little" ;                                                                                           
        double lat_bnds(lat, bnds) ;                                                                                                   
                lat_bnds:_Storage = "chunked" ;                                                                                        
                lat_bnds:_ChunkSizes = 180, 2 ;                                                                                        
                lat_bnds:_DeflateLevel = 1 ;                                                                                           
                lat_bnds:_Shuffle = "true" ;                                                                                           
                lat_bnds:_Endianness = "little" ;                                                                                      
        double lon(lon) ;                                                                                                              
...                                                                                     
                lon:_Storage = "contiguous" ;                                                                                          
                lon:_Endianness = "little" ;                                                                                           
        double lon_bnds(lon, bnds) ;                                                                                                   
                lon_bnds:_Storage = "chunked" ;                                                                                        
                lon_bnds:_ChunkSizes = 360, 2 ;                                                                                        
                lon_bnds:_DeflateLevel = 1 ;                                                                                           
                lon_bnds:_Shuffle = "true" ;                                                                                           
                lon_bnds:_Endianness = "little" ;                                                                                      
        float areacello(lat, lon) ;                                                                                                    
...
                areacello:_FillValue = 1.e+20f ;
                areacello:_Storage = "chunked" ;
                areacello:_ChunkSizes = 180, 360 ;
                areacello:_DeflateLevel = 1 ;
                areacello:_Shuffle = "true" ;
                areacello:_Endianness = "little" ;

// global attributes:
...
                :cmor_version = "3.2.2" ;
                :_NCProperties = "version=1|netcdflibversion=4.4.1|hdf5libversion=1.8.17" ;
                :_SuperblockVersion = 0 ;
                :_IsNetcdf4 = 1 ;
                :_Format = "netCDF-4 classic model" ;
}
```

**Comment by durack1** [comment]
@dnadeau4 ok so the [`superblockVersion` change is just due to a `netcdf 4.4.1+` update](https://github.com/Unidata/netcdf-c/releases/tag/v4.4.1), so the difference here is that the coords are no longer being compressed, when in `cmor 3.1.2` they were.. We're also [up-to-date with `libnetcdf 4.4.1.1` on conda-forge](https://anaconda.org/conda-forge/libnetcdf/files), so it's a `cmor`-specific issue I believe

**Comment by dnadeau4** [comment]
I change this code to not compress/chunk by default.    Only the `var_id` will be compressed.  There is no way to compress coordinate variables at this time,  only the variable.  I could create a cmor_get_varid that you could then pass to the the cmor_set_var_chunking function if needed.

**Comment by durack1** [comment]
@dnadeau4 did you expect the behavior to change? It seems from the ncdump output above, `3.1.2` WAS compressing coords, whereas that's no longer the case in `3.2.2`

**Comment by dnadeau4** [comment]
This was a request from UK.  They were not happy with the chunking setup in CMOR.   Deflation/chunking are set in the same function.

**Comment by dnadeau4** [comment]
https://github.com/PCMDI/cmor/issues/137

**Comment by durack1** [comment]
@ehogan @jkettleb did you guys want this different behavior regarding coord deflation? It seems like an inconsistency to me.. I would prefer that deflation returns to the same behavior that was in `3.1.2` as noted above

**Comment by durack1** [comment]
@dnadeau4 digging a little deeper, it seems that the `3.1.2` to `3.2.3` change has bloated the larger file sizes considerably (both `tos*` variables) but also reduced some file sizes by 30%!! (`siconc`):
**3.1.2**
```
[duro@ocean ~]$ ls -sh /work/durack1/Shared/150219_AMIPForcingData/CMIP6/input4MIPs/PCMDI/SSTsAndSeaIce/CMIP/*/*/PCMDI-AMIP-1-1-1/*/gn/v20161020/*.nc
 44K v20161020/areacello_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn.nc
No sftof in v1.1.1
 74M v20161020/siconcbcs_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn_187001-201606.nc
 45M v20161020/siconc_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn_187001-201606.nc
186M v20161020/tosbcs_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn_187001-201606.nc
178M v20161020/tos_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn_187001-201606.nc
```
**3.2.3**
```
[duro@ocean ~]$ ls -sh /work/durack1/Shared/150219_AMIPForcingData/CMIP6/input4MIPs/PCMDI/SSTsAndSeaIce/CMIP/*/*/PCMDI-AMIP-1-1-2/*/gn/v20170419/*.nc
 48K v20170419/areacello_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn.nc
 52K v20170419/sftof_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn.nc
 74M v20170419/siconcbcs_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn_187001-201612.nc
 30M v20170419/siconc_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn_187001-201612.nc
239M v20170419/tosbcs_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn_187001-201612.nc
231M v20170419/tos_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn_187001-201612.nc
```

**Comment by taylor13** [comment]
@durack1    I think the change in temperature units from kelvin to degC explains the "bloating" of the SST files.  The kelvin numbers like 27x.xxx, 28x.xxxx, 29x.xxxx have more digits in common than their degC counterparts (e.g., all numbers start with 2, whereas the degC numbers don't all have a common digit).

I can't explain why the sea ice concentration files have gotten smaller.  I think @dnadeau4 noted a huge reduction in chunk size specified for the time coordinate, which might explain some of it.  Also, perhaps the change from decimal fraction to percentage helps compression, but I don't see how.

**Comment by durack1** [comment]
@taylor13 @dnadeau4 here is the `siconc` output to compare. I think you might be onto something with the difference

3.2.3: `time:_ChunkSizes = 512 ;` and 3.1.2: `time:_ChunkSizes = 524288 ;`.

The ncdump output of the v3.1.2 file (45Mb):
[siconc_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn_187001-201606.txt](https://github.com/PCMDI/cmor/files/962765/siconc_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn_187001-201606.txt)
```
netcdf siconc_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-1_gn_187001-201606 {
dimensions:
	time = UNLIMITED ; // (1758 currently)
	lat = 180 ;
	lon = 360 ;
	bnds = 2 ;
variables:
	double time(time) ;
...
		time:_Storage = "chunked" ;
		time:_ChunkSizes = 524288 ;
		time:_DeflateLevel = 1 ;
		time:_Shuffle = "true" ;
		time:_Endianness = "little" ;
	double time_bnds(time, bnds) ;
		time_bnds:_Storage = "chunked" ;
		time_bnds:_ChunkSizes = 1, 2 ;
		time_bnds:_DeflateLevel = 1 ;
		time_bnds:_Shuffle = "true" ;
		time_bnds:_Endianness = "little" ;
	double lat(lat) ;
...
		lat:_Storage = "chunked" ;
		lat:_ChunkSizes = 180 ;
		lat:_DeflateLevel = 1 ;
		lat:_Shuffle = "true" ;
		lat:_Endianness = "little" ;
	double lat_bnds(lat, bnds) ;
		lat_bnds:_Storage = "chunked" ;
		lat_bnds:_ChunkSizes = 180, 2 ;
		lat_bnds:_DeflateLevel = 1 ;
		lat_bnds:_Shuffle = "true" ;
		lat_bnds:_Endianness = "little" ;
	double lon(lon) ;
...
		lon:_Storage = "chunked" ;
		lon:_ChunkSizes = 360 ;
		lon:_DeflateLevel = 1 ;
		lon:_Shuffle = "true" ;
		lon:_Endianness = "little" ;
	double lon_bnds(lon, bnds) ;
		lon_bnds:_Storage = "chunked" ;
		lon_bnds:_ChunkSizes = 360, 2 ;
		lon_bnds:_DeflateLevel = 1 ;
		lon_bnds:_Shuffle = "true" ;
		lon_bnds:_Endianness = "little" ;
	float siconc(time, lat, lon) ;
...
		siconc:units = "1.0" ;
		siconc:original_units = "%" ;
		siconc:history = "2016-10-20T19:52:04Z altered by CMOR: Converted units from \'%\' to \'1.0\'." ;
...		siconc:_Storage = "chunked" ;
		siconc:_ChunkSizes = 1, 180, 360 ;
		siconc:_DeflateLevel = 1 ;
		siconc:_Shuffle = "true" ;
		siconc:_Endianness = "little" ;

// global attributes:
...
		:cmor_version = "3.1.2" ;
		:_SuperblockVersion = 2 ;
		:_IsNetcdf4 = 1 ;
		:_Format = "netCDF-4 classic model" ;
}
```
The ncdump output of the v3.2.2 file (30Mb):
[siconc_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn_187001-201612.txt](https://github.com/PCMDI/cmor/files/962761/siconc_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn_187001-201612.txt)
```
netcdf siconc_input4MIPs_SSTsAndSeaIce_CMIP_PCMDI-AMIP-1-1-2_gn_187001-201612 {
dimensions:
	time = UNLIMITED ; // (1764 currently)
	lat = 180 ;
	lon = 360 ;
	bnds = 2 ;
variables:
	double time(time) ;
...
		time:_Storage = "chunked" ;
		time:_ChunkSizes = 512 ;
		time:_Endianness = "little" ;
	double time_bnds(time, bnds) ;
		time_bnds:_Storage = "chunked" ;
		time_bnds:_ChunkSizes = 1, 2 ;
		time_bnds:_DeflateLevel = 1 ;
		time_bnds:_Shuffle = "true" ;
		time_bnds:_Endianness = "little" ;
	double lat(lat) ;
...
		lat:_Storage = "contiguous" ;
		lat:_Endianness = "little" ;
	double lat_bnds(lat, bnds) ;
		lat_bnds:_Storage = "chunked" ;
		lat_bnds:_ChunkSizes = 180, 2 ;
		lat_bnds:_DeflateLevel = 1 ;
		lat_bnds:_Shuffle = "true" ;
		lat_bnds:_Endianness = "little" ;
	double lon(lon) ;
...
		lon:_Storage = "contiguous" ;
		lon:_Endianness = "little" ;
	double lon_bnds(lon, bnds) ;
		lon_bnds:_Storage = "chunked" ;
		lon_bnds:_ChunkSizes = 360, 2 ;
		lon_bnds:_DeflateLevel = 1 ;
		lon_bnds:_Shuffle = "true" ;
		lon_bnds:_Endianness = "little" ;
	float siconc(time, lat, lon) ;
...
		siconc:units = "%" ;
...
		siconc:_FillValue = 1.e+20f ;
		siconc:_Storage = "chunked" ;
		siconc:_ChunkSizes = 1, 180, 360 ;
		siconc:_DeflateLevel = 1 ;
		siconc:_Shuffle = "true" ;
		siconc:_Endianness = "little" ;

// global attributes:
...
		:cmor_version = "3.2.3" ;
		:_NCProperties = "version=1|netcdflibversion=4.4.1.1|hdf5libversion=1.8.17" ;
		:_SuperblockVersion = 0 ;
		:_IsNetcdf4 = 1 ;
		:_Format = "netCDF-4 classic model" ;
}
```

**Comment by ehogan** [comment]
> @ehogan @jkettleb did you guys want this different behavior regarding coord deflation? It seems like an inconsistency to me.. I would prefer that deflation returns to the same behavior that was in 3.1.2 as noted above

@durack1 apologies for the delay in my reply; I returned to the office yesterday after 2 weeks of leave :)

The original issue I submitted about the chunking is #130. The chunking was introduced in CMOR 3.1.3 and reverted in CMOR 3.2.2. I confirmed the data I was using for testing behaved in the same way for CMOR 3.2.2 as it did in CMOR 3.1.2. 

Is the deflation issue not related to the fact that the netCDF version is different in your examples, as you mentioned [here](https://github.com/PCMDI/cmor/issues/158#issuecomment-294379040)? I didn't notice any changes or issues with the deflation level during my testing and I'm using netCDF 4.3.1 (regardless of CMOR version; I'm using a local, non-conda installation).

**Comment by ehogan** [comment]
> Is the deflation issue not related to the fact that the netCDF version is different in your examples, as you mentioned here? 

I misinterpreted your comment, so I did some more digging:

Even though the behaviour of CMOR 3.2.2 is the same as CMOR 3.1.2, I didn't notice that the `i` and `j` variables in the output netCDF files from CMOR 3.2.2 are missing the `_ChunkSizes` and `_DeflateLevel` attributes. My apologies for not spotting this before.

I made the assumption that the changes made to CMOR 3.2.2 enabled the chunking to be applied in the same way as it was in CMOR 3.1.2, rather than removing chunking completely for particular variables. It makes sense to apply chunking / deflation consistently to all variables in the output netCDF files, as long as the current behaviour doesn't change (as described in #130).

**Comment by dnadeau4** [comment]
Can you retry in 3.2.4?  I fix part of it, although the `time` dimension is still problematic.  Will fix in `3.2.5`

**Comment by dnadeau4** [comment]
@durack1 I default the chunking to `time` `coordinate_variable` to `1`, unless the user overwrite it.  

other `coordinate_variables` are set to `CONTIGUOUS` (No Chunking) by default.

main variable `(ta for example)` is chunked  to `1, 1, len(Y), len(X)` creating a 2D tile data array.

**Comment by dnadeau4** [comment]
Fix in 3.2.5 Please @doutriaux1 @ehogan try and reopen if needed.
